### PR TITLE
Add upgradeability utils

### PR DESCRIPTION
**Pull request by webthethird**
Extends #1699, adding more utilities to `slither.utils.upgradeability`, specifically related to finding a proxy's implementation address storage slot.

### CODE REVIEW

**Comment by webthethird** [comment]
I'm not sure why `Test slither-read-storage` is failing here, since this PR doesn't touch it. I checked the test details, here's what it says:
```
Run pytest tests/test_read_storage.py
bash: cannot set terminal process group (669): Inappropriate ioctl for device
bash: no job control in this shell
============================= test session starts ==============================
platform linux -- Python 3.8.16, pytest-[7](https://github.com/crytic/slither/actions/runs/4418537918/jobs/7745793871#step:7:8).2.2, pluggy-1.0.0
rootdir: /home/runner/work/slither/slither
plugins: cov-4.0.0, web3-6.0.0, xdist-3.2.1
collected 0 items / 1 error

==================================== ERRORS ====================================
_________________ ERROR collecting tests/test_read_storage.py __________________
slither/tools/read_storage/read_storage.py:9: in <module>
    from eth_abi import decode_single, encode_abi
E   ImportError: cannot import name 'decode_single' from 'eth_abi' (/opt/hostedtoolcache/Python/3.[8](https://github.com/crytic/slither/actions/runs/4418537918/jobs/7745793871#step:7:9).16/x64/lib/python3.8/site-packages/eth_abi/__init__.py)

During handling of the above exception, another exception occurred:
tests/test_read_storage.py:13: in <module>
    from slither.tools.read_storage import SlitherReadStorage
slither/tools/read_storage/__init__.py:1: in <module>
    from .read_storage import SlitherReadStorage
slither/tools/read_storage/read_storage.py:1[9](https://github.com/crytic/slither/actions/runs/4418537918/jobs/7745793871#step:7:10): in <module>
    sys.exit(-1)
E   SystemExit: -1
------------------------------- Captured stdout --------------------------------
ERROR: in order to use slither-read-storage, you need to install web3
$ pip3 install web3 --user

=========================== short test summary info ============================
ERROR tests/test_read_storage.py - SystemExit: -1
!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!
=============================== 1 error in 0.61s ===============================
Error: Process completed with exit code 2.
```
Seems unrelated to me, though I was hoping the problem would go away when I merged dev into this PR.

File: slither/utils/upgradeability.py [hunk 1]
```diff
@@ -0,0 +1,463 @@
...
+    TemporaryVariable,
+    TupleVariable,
+    Constant,
+    ReferenceVariable,
+)
+from slither.tools.read_storage.read_storage import SlotInfo, SlitherReadStorage
+
+
+# pylint: disable=too-many-locals
+def compare(v1: Contract, v2: Contract) -> dict:
```
Comment (montyly) [inline slither/utils/upgradeability.py:63]: Should we return a dict here? Maybe we can return the 6 list directly?

Using a dict will make it difficult to use or to use python types

File: slither/utils/upgradeability.py [hunk 2]
```diff
@@ -0,0 +1,463 @@
...
+        "missing-vars-in-v2": list[Variable],
+        "new-variables": list[Variable],
+        "tainted-variables": list[Variable],
+        "new-functions": list[Function],
+        "modified-functions": list[Function],
+        "tainted-functions": list[Function]
+    }
+    """
+
+    order_vars1 = [v for v in v1.state_variables if not v.is_constant and not v.is_immutable]
```
Comment (montyly) [inline slither/utils/upgradeability.py:82]: Here we want to use `state_variables_ordered` https://github.com/crytic/slither/blob/master/slither/core/declarations/contract.py#L363

`state_variables` does not maintain the order of declarations, and dont include private variables:
https://github.com/crytic/slither/blob/64b2e0c063cf93fb479481abc432371404629f32/slither/core/declarations/contract.py#L345-L349

File: slither/utils/upgradeability.py [hunk 3]
```diff
@@ -0,0 +1,463 @@
...
+        "missing-vars-in-v2": [],
+        "new-variables": [],
+        "tainted-variables": [],
+        "new-functions": [],
+        "modified-functions": [],
+        "tainted-functions": [],
+    }
+
+    # Since this is not a detector, include any missing variables in the v2 contract
+    if len(order_vars2) < len(order_vars1):
```
Comment (montyly) [inline slither/utils/upgradeability.py:97]: It looks like we are dupplicating existing logic. Ex: https://github.com/crytic/slither/blob/master/slither/tools/upgradeability/checks/variables_order.py#L51-L68

I think here we should:
- Create utils.upgradeability based on the upgradeability detectors 
- Use the utils in the upgradeability detectors

File: slither/utils/upgradeability.py [hunk 4]
```diff
@@ -0,0 +1,463 @@
...
+    for sig in func_sigs2:
+        function = v2.get_function_from_signature(sig)
+        orig_function = v1.get_function_from_signature(sig)
+        if sig not in func_sigs1:
+            new_modified_functions.append(function)
+            results["new-functions"].append(function)
+            new_modified_function_vars += (
+                function.state_variables_read + function.state_variables_written
+            )
+        elif not function.name.startswith("slither") and is_function_modified(
```
Comment (montyly) [inline slither/utils/upgradeability.py:114]: I think here we should use `not function. is_constructor_variables` instead of comparing the name

https://github.com/crytic/slither/blob/master/slither/core/declarations/function.py#L396-L404

File: slither/utils/upgradeability.py [hunk 5]
```diff
@@ -0,0 +1,463 @@
...
+            for var in set(new_modified_function_vars)
+            if var in function.variables_read_or_written
+            and not var.is_constant
+            and not var.is_immutable
+        ]
+        if len(modified_calls) > 0 or len(tainted_vars) > 0:
+            results["tainted-functions"].append(function)
+
+    # Find all new or tainted variables, i.e., variables that are read or written by a new/modified/tainted function
+    for _, var in enumerate(order_vars2):
```
Comment (montyly) [inline slither/utils/upgradeability.py:146]: Do we need `enumerate` here? It looks like can do `for var in order_vars2` ?

File: slither/utils/upgradeability.py [hunk 6]
```diff
@@ -0,0 +1,463 @@
...
+    Gets the Variable that stores a proxy's implementation address. Uses data dependency to trace any LocalVariable
+    that is passed into a delegatecall as the target address back to its data source, ideally a StateVariable.
+    Args:
+        proxy: A Contract object (proxy.is_upgradeable_proxy should be true).
+
+    Returns:
+        (`Variable`) | None : The variable, ideally a StateVariable, which stores the proxy's implementation address.
+    """
+    available_functions = proxy.available_functions_as_dict()
+    if not proxy.is_upgradeable_proxy or not available_functions["fallback()"]:
```
Comment (montyly) [inline slither/utils/upgradeability.py:344]: I think this can lead to issues if a function is named `fallback()`

We should iterate over the function, and look for `function. is_fallback`

File: slither/utils/upgradeability.py [hunk 7]
```diff
@@ -0,0 +1,463 @@
...
+    Searches a proxy's fallback function for a delegatecall, then extracts the Variable being passed in as the target.
+    Should typically be called by get_proxy_implementation_var(proxy).
+    Args:
+        proxy: A Contract object (should have a fallback function).
+
+    Returns:
+        (`Variable`) | None : The variable being passed as the destination argument in a delegatecall in the fallback.
+    """
+    delegate: Optional[Variable] = None
+    fallback = proxy.available_functions_as_dict()["fallback()"]
```
Comment (montyly) [inline slither/utils/upgradeability.py:368]: Same as before, we should use `is_fallback`

File: slither/utils/upgradeability.py [hunk 8]
```diff
@@ -0,0 +1,463 @@
...
+                and len(expression.arguments) > 1
+            ):
+                dest = expression.arguments[1]
+                if isinstance(dest, CallExpression) and "sload" in str(dest.called):
+                    dest = dest.arguments[0]
+                if isinstance(dest, Identifier):
+                    delegate = dest.value
+                    break
+                if isinstance(dest, Literal) and len(dest.value) == 66:
+                    delegate = StateVariable()
```
Comment (montyly) [inline slither/utils/upgradeability.py:397]: That's an interesting case that we dont usually use: having a StateVariable created on the fly by an analysis.

This might break assumption in other part of the codebase, we should probably document that `find_delegate_in_fallback` can return a newly created variable

File: slither/utils/upgradeability.py [hunk 9]
```diff
@@ -0,0 +1,463 @@
...
+                and "delegatecall" in str(expression.called)
+                and len(expression.arguments) > 1
+            ):
+                dest = expression.arguments[1]
+                if isinstance(dest, CallExpression) and "sload" in str(dest.called):
+                    dest = dest.arguments[0]
+                if isinstance(dest, Identifier):
+                    delegate = dest.value
+                    break
+                if isinstance(dest, Literal) and len(dest.value) == 66:
```
Comment (montyly) [inline slither/utils/upgradeability.py:396]: can you comment the magic value?

File: slither/utils/upgradeability.py [hunk 10]
```diff
@@ -0,0 +1,463 @@
...
+                    delegate = StateVariable()
+                    delegate.is_constant = True
+                    delegate.expression = dest
+                    delegate.name = dest.value
+                    delegate.type = ElementaryType("bytes32")
+                    break
+    return delegate
+
+
+def extract_delegate_from_asm(contract: Contract, node: Node) -> Optional[Variable]:
```
Comment (montyly) [inline slither/utils/upgradeability.py:406]: This function needs extra testing due to the manual parsing

File: slither/utils/upgradeability.py [hunk 11]
```diff
@@ -0,0 +1,463 @@
...
+    Returns:
+        (`Variable`) | None : The variable with the matching name, if found
+    """
+    for sv in contract.state_variables:
+        if sv.name == dest:
+            return sv
+    for lv in parent_func.local_variables:
+        if lv.name == dest:
+            return lv
+    for pv in parent_func.parameters:
```
Comment (montyly) [inline slither/utils/upgradeability.py:460]: maybe `parent_func.returns` too? In case the variable is declared in the return statement

**Review by montyly — COMMENTED** [review]
Great work

- Can you run mypy on the upgradeability.py?  We are in the process of adding it to the CI. You can use the same setup than for crytic-compile (https://github.com/crytic/crytic-compile/blob/master/CONTRIBUTING.md#linters)
- I think we need further testing. We should have at least 1 test per function in `upgradeability`. The comparison using a json is probably not something robust enough for here. (cc @0xalpharush )

File: slither/utils/upgradeability.py (hunk 3 above)
Comment (webthethird) [reply slither/utils/upgradeability.py:104]: My latest commit makes `get_missing_vars` its own utility function and uses it in the MissingVariable upgradeability detector.

File: slither/utils/upgradeability.py (hunk 10 above)
Comment (webthethird) [reply slither/utils/upgradeability.py:443]: >This function needs extra testing due to the manual parsing

This function is mostly used to handle contracts with solidity versions < 0.6.0. I chose three older proxies to test this with, and updated the utility function based on my testing. But one of the three, which has the most common structure, failed my test due to issue #1775. For more extensive testing I suspect I would need a work-around for that issue.

**Comment by montyly** [comment]
Outside of the conflict with `dev`, is this ready for another review?

**Comment by webthethird** [comment]
> Outside of the conflict with `dev`, is this ready for another review?

I believe so

File: slither/utils/upgradeability.py [hunk 12]
```diff
@@ -0,0 +1,568 @@
...
+    order_vars1 = [
+        v for v in v1.state_variables_ordered if not v.is_constant and not v.is_immutable
+    ]
+    order_vars2 = [
+        v for v in v2.state_variables_ordered if not v.is_constant and not v.is_immutable
+    ]
+    func_sigs1 = [function.solidity_signature for function in v1.functions]
+    func_sigs2 = [function.solidity_signature for function in v2.functions]
+
+    results = {
```
Comment (montyly) [inline slither/utils/upgradeability.py:94]: I think we can remove the dict here, and directly use the variables. It will reduce the variables complexity and help the types checker

File: slither/utils/upgradeability.py [hunk 13]
```diff
@@ -0,0 +1,568 @@
...
+        visited.extend([node_f1, node_f2])
+        queue_f1.extend(son for son in node_f1.sons if son not in visited)
+        queue_f2.extend(son for son in node_f2.sons if son not in visited)
+        for i, ir in enumerate(node_f1.irs):
+            if encode_ir_for_compare(ir) != encode_ir_for_compare(node_f2.irs[i]):
+                return True
+    return False
+
+
+def ntype(_type):  # pylint: disable=too-many-branches
```
Comment (montyly) [inline slither/utils/upgradeability.py:229]: Add function signature type

File: slither/utils/upgradeability.py [hunk 14]
```diff
@@ -0,0 +1,568 @@
...
+    if "mapping" in _type:
+        return "mapping"
+    return _type.replace(" ", "_")
+
+
+def encode_ir_for_compare(ir) -> str:  # pylint: disable=too-many-branches
+    # operations
+    if isinstance(ir, Assignment):
+        return f"({encode_ir_for_compare(ir.lvalue)}):=({encode_ir_for_compare(ir.rvalue)})"
+    if isinstance(ir, Index):
```
Comment (montyly) [inline slither/utils/upgradeability.py:268]: Add: `ir: Operation`

**Review by montyly — COMMENTED** [review]
This looks good. Only minor comment to addressed, and we can merge it

**Comment by webthethird** [comment]
@montyly I believe I've addressed the minor changes

File: slither/utils/upgradeability.py [hunk 15]
```diff
@@ -0,0 +1,570 @@
...
+        return "tuple"
+    if "contract" in _type:
+        return "contract"
+    if "mapping" in _type:
+        return "mapping"
+    return _type.replace(" ", "_")
+
+
+# pylint: disable=too-many-branches
+def encode_ir_for_compare(ir: Union[Operation, Variable]) -> str:
```
Comment (montyly) [inline slither/utils/upgradeability.py:266]: It's great that we have added the types, because this does not sound right :) 

Should we split this function in two?
- `encore_ir_for_compare`
- `encore_variable_for_compare`

Note that we probably have the same issue in the slither-similar code, but that should be fixed in another PR

Comment (webthethird) [reply slither/utils/upgradeability.py]: >Should we split this function in two?

Done!

**Comment by montyly** [comment]
@0xalpharush : should I merge this, or do you want to take a stab at the tests architecture first?

File: slither/core/declarations/contract.py [hunk 16]
```diff
@@ -663,6 +666,24 @@ def functions_and_modifiers_declared(self) -> List["Function"]:
         """
         return self.functions_declared + self.modifiers_declared  # type: ignore
 
+    @property
```
Comment (0xalpharush) [inline slither/core/declarations/contract.py:669]: Can we add a test case under `slither/tests/unit/core` to ensure that it works for examples like https://github.com/crytic/slither/issues/1331#issuecomment-1209433713

**Comment by 0xalpharush** [comment]
@montyly I think we should merge `dev` into this branch so that the tests are run in CI

File: slither/core/declarations/contract.py (hunk 16 above)
Comment (webthethird) [reply slither/core/declarations/contract.py:669]: Done!

File: tests/test_upgradeability_util.py
[the remark this answers is not in the document]
Comment (webthethird) [reply tests/test_upgradeability_util.py:1]: Done

**Comment by webthethird** [comment]
> I think we should merge `dev` into this branch so that the tests are run in CI

Done, done and done!
